mlflow/store/analytics/trace_correlation.py

```diff
@@ -0,0 +1,249 @@
+    - -1: Perfect negative correlation (events never co-occur)
+    - 0: Independence (events occur independently)
+    - 1: Perfect positive correlation (events always co-occur)
+    - NaN: Undefined (when one or both events have zero count)
+
+    Args:
+        joint_count: Number of times both events occur together
+        filter1_count: Number of times event 1 occurs
+        filter2_count: Number of times event 2 occurs
+        total_count: Total number of observations
+
+    Returns:
+        NPMIResult containing both unsmoothed and smoothed NPMI values.
+
+    Examples:
+        >>> result = calculate_npmi_from_counts(10, 20, 15, 100)
+        >>> result.npmi  # Unsmoothed value
+        >>> result.npmi_smoothed  # Smoothed value
+    """
+    # No population
+    if total_count <= 0:
+        return NPMIResult(npmi=float("nan"), npmi_smoothed=float("nan"))
+
+    # Return NaN if either filter has zero support
+    if filter1_count == 0 or filter2_count == 0:
+        return NPMIResult(npmi=float("nan"), npmi_smoothed=float("nan"))
+
+    n11 = joint_count  # Both occur
+    n10 = filter1_count - joint_count  # Only filter1
+    n01 = filter2_count - joint_count  # Only filter2
+    n00 = total_count - filter1_count - filter2_count + joint_count  # Neither
+
+    if min(n11, n10, n01, n00) < 0:
+        # Inconsistent counts, return undefined
+        return NPMIResult(npmi=float("nan"), npmi_smoothed=float("nan"))
+
+    # Calculate unsmoothed NPMI with explicit -1.0 rule
+    if joint_count == 0 and filter1_count > 0 and filter2_count > 0:
+        npmi_unsmoothed = -1.0
+    else:
+        npmi_unsmoothed = _calculate_npmi_core(n11, n10, n01, n00, smoothing=0)
+
+    # Calculate smoothed NPMI for robustness
+    npmi_smoothed = _calculate_npmi_core(n11, n10, n01, n00, smoothing=JEFFREYS_PRIOR)
+
+    return NPMIResult(npmi=npmi_unsmoothed, npmi_smoothed=npmi_smoothed)
+
+
+def _calculate_npmi_core(
+    n11: float,
+    n10: float,
+    n01: float,
+    n00: float,
+    smoothing: float = 0,
+) -> float:
+    """
+    Core NPMI calculation with optional smoothing.
+
+    Internal function that performs the actual NPMI calculation
+    on a 2x2 contingency table with optional additive smoothing.
+
+    Args:
+        n11: Count of both events occurring
+        n10: Count of only event 1 occurring
+        n01: Count of only event 2 occurring
+        n00: Count of neither event occurring
+        smoothing: Additive smoothing parameter (0 for no smoothing)
+
+    Returns:
+        NPMI value in [-1, 1], or NaN if undefined.
+    """
+    n11_s = n11 + smoothing
+    n10_s = n10 + smoothing
+    n01_s = n01 + smoothing
+    n00_s = n00 + smoothing
+
+    N = n11_s + n10_s + n01_s + n00_s
+    n1 = n11_s + n10_s  # Total event 1 count
+    n2 = n11_s + n01_s  # Total event 2 count
+
+    # Check for zero marginals (shouldn't happen with positive smoothing)
+    if n1 <= 0 or n2 <= 0 or n11_s <= 0:
+        return float("nan")
```

Comment:
The prototype had a forcing cast to -1 for this case, but I don't believe this is mathematically correct. NaN can be handled by our RPC, so I figured it might be better to pass the true result of this edge case. 
If you really don't want to do this, let me know, but I think having either this or a sentinel value added to indicate that the value being passed is actually invalid is required to make this accurate.